### ISSUE TITLE
Diagnostics get out of sync

### ISSUE THREAD

**Issue by rchl** [sublimelsp/LSP#1198]
During editing of JSON file the diagnostics become out of sync with the real state of the file. Happened before in other servers and files also.

```
	"diagnostics_delay_ms": 2000,
	"diagnostics_additional_delay_auto_complete_ms": 1000,
```
`st4000-exploration` (b1a1fa79a71d215dde2d6c396b559561feaeef5f).

```
::  -> LSP-json textDocument/didChange: {'textDocument': {'uri': 'file:///project/template/.lokalise.json', 'version': 6}, 'contentChanges': [{'text': '        "filter_langs": ["bg", "de", "en", "es", "fr", "it", "no", "pl", "pt", "pt_BR"],\n', 'range': {'end': {'character': 0, 'line': 7}, 'start': {'character': 0, 'line': 7}}, 'rangeLength': 0}]}
:: <-  LSP-json textDocument/publishDiagnostics: {'diagnostics': [{'message': 'Duplicate object key', 'source': 'json', 'code': 520, 'range': {'end': {'character': 22, 'line': 8}, 'start': {'character': 8, 'line': 8}}, 'severity': 2}, {'message': 'Duplicate object key', 'source': 'json', 'code': 520, 'range': {'end': {'character': 22, 'line': 7}, 'start': {'character': 8, 'line': 7}}, 'severity': 2}], 'uri': 'file:///project/template/.lokalise.json'}
::  -> LSP-json textDocument/didChange: {'textDocument': {'uri': 'file:///project/template/.lokalise.json', 'version': 7}, 'contentChanges': [{'text': '', 'range': {'end': {'character': 0, 'line': 9}, 'start': {'character': 0, 'line': 8}}, 'rangeLength': 86}]}
:: --> LSP-json textDocument/documentColor(10): {'textDocument': {'uri': 'file:///project/template/.lokalise.json'}}
:: <<< LSP-json 10: []
:: <-  LSP-json textDocument/publishDiagnostics: {'diagnostics': [], 'uri': 'file:///project/template/.lokalise.json'}
```

![Screenshot 2020-07-20 at 12 53 10](https://user-images.githubusercontent.com/153197/87930226-f94a1900-ca87-11ea-9792-d52aa9137e18.png)

As clear from the log, the last state from the server reported no diagnostics but ST is still showing the one that was reported before that.

**Comment by ayoub-benali** [comment]
Iam having a similar issue with LSP-metals

**Comment by rchl** [comment]
Are you also using non-default value for `diagnostics_delay_ms`?

**Comment by ayoub-benali** [comment]
> Are you also using non-default value for `diagnostics_delay_ms`?

No I am using default values for both settings.

**Comment by ayoub-benali** [comment]
LSP isn't showing `textDocument/publishDiagnostics`, sometime the panel gets updated after compilation sometime not.

These are my settings: 
```json
  "auto_show_diagnostics_panel": "never",
  "show_diagnostics_in_view_status": false,
  "show_diagnostics_count_in_view_status": true,
  "diagnostics_panel_include_severity_level": 1,
```
in the server logs I see clearly: 
```
:: <-  LSP-metals textDocument/publishDiagnostics: {'diagnostics': [{'severity': 1,
```


I can create a separate issue if you think it isn't related.

**Comment by rwols** [comment]
Is there a `"version"` present in the diagnostics payload? I have a hunch that the problem is related to the diagnostics version.

**Comment by ayoub-benali** [comment]
Not in my case.
I also noticed that the counter in the status Bar and the diagnostic panel are out sync. Meaning Errors count is 0 but diagnostic panel still reports an error.

**Comment by rwols** [comment]
> I also noticed that the counter in the status Bar and the diagnostic panel are out sync. Meaning Errors count is 0 but diagnostic panel still reports an error.

The diagnostics count in the status bar shows the count of the current file, the panel shows all diagnostics. Are the files the same?

**Comment by rwols** [comment]
Oh, I realize now that that status bar count was a global error count, not per-view. https://github.com/sublimelsp/LSP/blob/e7dac40247d68398fc24d1b19b1565d05e472ae8/plugin/diagnostics.py#L301

**Comment by ayoub-benali** [comment]
> Oh, I realize now that that status bar count was a global error count, not per-view.

I think it was always a global count. I often check it before deciding to open or not the diagnostic panel

**Comment by rwols** [comment]
I've created an issue for it #1201

**Comment by rwols** [comment]
The discussion in https://github.com/sublimelsp/LSP/pull/1203 shows that there is a problem with debounced calls not being sequential (debouncing f at time t1 and debouncing f at time t2 might result in the second f being called earlier, if that makes sense).

But @ayoub-benali notes that his `diagnostics_delay_ms` setting is zero, so that puzzles me a bit.

**Comment by rchl** [comment]
> debouncing f at time t1 and debouncing f at time t2 might result in the second f being called earlier, if that makes sense

The two denounced calls would be sequential. In that particular case second call is not denounced so it runs before the first one that was debounced.

**Comment by ayoub-benali** [comment]
So even after https://github.com/sublimelsp/LSP/pull/1202 was merged I still have the same issue, although it seems random and I can't reproduce the issue easily. 
The error count and diagnostic panel are respectively 0 and empty but looking at the server logs I see 
`:: <-  LSP-metals textDocument/publishDiagnostics: {'diagnostics'` with severity 1

So both issues have probably the same source.

**Comment by rchl** [comment]
Note that the fix for the issue I've reported has not been made yet.

**Comment by rchl** [comment]
I think I found the other issue apart from the one caused by denounced functions.

It appears to be an ST issue with regions updating.
On "undoing" a change, ST also restores old diagnostic, if there was one. If language server won't send new `publishDiagnostics` (because based on document state it's not necessary) then we'll be left with that old diagnostic in view.

We could erase diagnostics more aggressively to fix that (which is probably what was happening before refactor) but I'm not sure what's the best approach here.

**Comment by rwols** [comment]
> The error count and diagnostic panel are respectively 0 and empty but looking at the server logs I see
:: <- LSP-metals textDocument/publishDiagnostics: {'diagnostics' with severity 1

Note that previously, diagnostics for files that were not opened were also handled (files that did not have a textDocument/didOpen). But now we don't do this anymore. Only files that are tracked receive diagnostics. Because it wasn't clear to me when these "viewless" diagnostics would be cleared from the internal storage. I could only conclude that we had an accidental memory leak. It was also problematic to key diagnostics by their filename, because the files `~/foo/bar.py` and `~/bar.p` could be the same file via symlinks. And `C:/FOO.PY` is the same file as `c:/foo.py` on Windows. The problem is somewhat alleviated with `os.path.samefile` which we use in at least one place. This function does an actual underlying file-system check instead of comparing strings. We may need a more general Path abstraction.

**Comment by rchl** [comment]
The second issue I can also only reproduce with `diagnostics_delay_ms` set to non-0 value.
That problem happens because our model is not necessarily aligned with the view since on "undo", ST re-adds regions and our model doesn't have any (until server publishes new diagnostics).

**Comment by rwols** [comment]
This should be fixed in 4070-1.0.6. Please open another issue for ST's undo behavior if it's important.
